src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,448 @@
+
+                if (link) {
+                        r = json_build(&v, JSON_BUILD_OBJECT(
+                                                JSON_BUILD_PAIR_STRING("TYPE", "MAP")));
+
+                        r = json_variant_merge_objectb(&v, JSON_BUILD_OBJECT(
+                                                JSON_BUILD_PAIR_OBJECT("DATA",
+                                                JSON_BUILD_PAIR_STRING("NAME", map_info.name),
+                                                JSON_BUILD_PAIR_UNSIGNED("TYPE", map_info.type),
+                                                JSON_BUILD_PAIR_STRING("MEMLOCK", memlock))));
+
+                        varlink_notify(link, v);
+                } else {
+                        log_struct(LOG_INFO,
+                                LOG_MESSAGE("Logging info about BPF map: %s", map_info.name),
+                                "MESSAGE_ID=" SD_MESSAGE_BPFD_LOG_STR,
+                                "NAME=%s", map_info.name,
+                                "TYPE=%u", map_info.type,
+                                "MEMLOCK=%s", memlock);
+                }
+        }
+
+        /* Iterate through all attached BPF progs*/
+        id = 0;
+        while (sym_bpf_prog_get_next_id(id, &id) >= 0) {
+                _cleanup_close_ int fd = -EBADF;
+                _cleanup_free_ char *memlock = NULL;
+                _cleanup_free_ uint32_t *prog_map_ids = NULL;
+                _cleanup_free_ char *prog_name = NULL;
+                _cleanup_free_ char *prog_map_names = NULL;
+                JsonVariant *v = NULL;
+
+                zero(prog_info);
+                /* Grab the FDs using their IDs*/
+                fd = sym_bpf_prog_get_fd_by_id(id);
+                if (fd < 0) {
+                        log_error("Can't get fd for prog id: %u", id);
+                        continue;
+                }
+
+                /* Fetch bpf_prog_info using fd */
+                r = sym_bpf_obj_get_info_by_fd(fd, &prog_info, &prog_info_len);
+                if (r < 0) {
+                        log_error("Can't get bpf prog info for fd: %i", fd);
+                }
+
+                /* If there are maps associated with this prog, fetch bpf_prog_info again
+                 * to get a list of map ids */
+                nr_map_ids = prog_info.nr_map_ids;
+                if (nr_map_ids > 0) {
+                        prog_map_ids = malloc(nr_map_ids * sizeof(uint32_t));
+                        if (!prog_map_ids)
+                                return -ENOMEM;
+                        prog_map_names = malloc(nr_map_ids * BPF_OBJ_NAME_LEN);
+                        if (!prog_map_names)
+                                return -ENOMEM;
+
+                        zero(prog_info);
+                        prog_info.nr_map_ids = nr_map_ids;
+                        prog_info.map_ids = PTR_TO_UINT64(prog_map_ids);
```

Comment:
we usually prefer structured initialization (except for very few cases where this doesn't work because union semantics are weird). ii.e.

```c
prog_info = (struct bpf_prog_info) {
        .nr_map_ids = nr_map_ids,
        .map_ids = PTR_TO_UINT64(prog_map_ids),
};
```